File: src/frontend/api/deduplicate.ts
import { apiClient } from './client'
import type { Media } from './types'

export interface DeduplicateGroup {
  clusterId: number
  items: Media[]
}

export interface DeduplicateGroupsRequest {
  cursor: string | null
  limit: number
}

export interface DeduplicateGroupsResponse {
  groups: DeduplicateGroup[]
  nextCursor: string | null
  hasMore: boolean
  totalGroups: number
  totalMedia: number
}

export interface DeduplicateActionResponse {
  message: string
  status: string
}

export interface DeduplicateStatusResponse {
  status: string
  runId: number | null
  trigger: string | null
  scheduledFor: string | null
  startedAt: string | null
  completedAt: string | null
  indexedMedia: number
  processedMedia: number
  candidateComparisons: number
  clustersCreated: number
  error: string | null
  nextScheduledAt: string | null
}

export const deduplicateApi = {
  groups: async (request: DeduplicateGroupsRequest): Promise<DeduplicateGroupsResponse> => {
    const response = await apiClient.post<DeduplicateGroupsResponse>('/ai/deduplicate/groups', request)
    return response.data
  },

  start: async (): Promise<DeduplicateActionResponse> => {
    const response = await apiClient.post<DeduplicateActionResponse>('/ai/deduplicate/start', {})
    return response.data
  },

  status: async (): Promise<DeduplicateStatusResponse> => {
    const response = await apiClient.post<DeduplicateStatusResponse>('/ai/deduplicate/status', {})
    return response.data
  },

  cancel: async (): Promise<DeduplicateActionResponse> => {
    const response = await apiClient.post<DeduplicateActionResponse>('/ai/deduplicate/cancel', {})
    return response.data
  },

  clean: async (): Promise<DeduplicateActionResponse> => {
    const response = await apiClient.post<DeduplicateActionResponse>('/ai/deduplicate/clean', {})
    return response.data
  },
}

File: src/backend/models/deduplicate.rs
use serde::{Deserialize, Serialize};

use crate::models::MediaResponse;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeduplicateGroupsRequest {
    pub cursor: Option<String>,
    pub limit: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeduplicateGroup {
    pub cluster_id: i64,
    pub items: Vec<MediaResponse>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeduplicateGroupsResponse {
    pub groups: Vec<DeduplicateGroup>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeduplicateActionResponse {
    pub message: String,
    pub status: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeduplicateStatusResponse {
    pub status: String,
    pub run_id: Option<i64>,
    pub trigger: Option<String>,
    pub scheduled_for: Option<String>,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
    pub indexed_media: i64,
    pub processed_media: i64,
    pub candidate_comparisons: i64,
    pub clusters_created: i64,
    pub error: Option<String>,
    pub next_scheduled_at: Option<String>,
}
